lib/util/trackmatcher.js

```diff
@@ -0,0 +1,149 @@
+'use strict';
+
+// NOTE(mroberts): TrackMatcher is meant to solve the problem identified in
+//
+//   https://bugs.webkit.org/show_bug.cgi?id=174519
+//
+// Namely that, without MIDs, we cannot "correctly" identify MediaStreamTracks
+// in Safari's current WebRTC implementation. So, this module tries to hack
+// around this by making a possibly dangerous assumption: "track" events will
+// be raised for MediaStreamTracks of a particular kind in the same order that
+// those kinds' MSIDs appear in the SDP. By calling `update` with an
+// RTCPeerConnection's `localDescription` and then invoking `match`, we ought
+// to be able to dequeue MediaStreamTrack IDs in the correct order to be
+// assigned to "track" events.
+
+/**
+ * @typedef {string} ID
+ */
+
+/**
+ * @interface MatchedAndUnmatched
+ * @property {Set<ID>} matched
+ * @property {Set<ID>} unmatched
+ */
+
+/**
+ * Create a new instance of {@link MatchedAndUnmatched}.
+ * @returns {MatchedAndUnmatched}
+ */
+function create() {
+  return {
+    matched: new Set(),
+    unmatched: new Set()
+  };
+}
+
+/**
+ * Attempt to match a MediaStreamTrack ID.
+ * @param {MatchedAndUnmatched} mAndM
+ * @returns {?string} id
+ */
+function match(mAndM) {
+  var unmatched = Array.from(mAndM.unmatched);
+  if (!unmatched.length) {
+    return null;
+  }
+  var id = unmatched[0];
+  mAndM.matched.add(id);
+  mAndM.unmatched = new Set(unmatched.slice(1));
+  return id;
+}
+
+/**
+ * Update a {@link MatchedAndUnmatched}'s MediaStreamTrack IDs.
+ * @param {MatchedAndUnmatched} mAndM
+ * @param {Set<ID>} ids
+ * @returns {void}
+ */
+function update(mAndM, ids) {
+  ids = new Set(ids);
+  mAndM.matched.forEach(function(id) {
+    if (!ids.has(id)) {
+      mAndM.matched.delete(id);
+    }
+  });
+  ids.forEach(function(id) {
```

Comment:
@markandrus This can be:
```javascript
var unmatchedIdsToAdd = util.difference(ids, mAndM.unmatched);
unmatchedIdsToAdd.forEach(mAndM.unmatched.add, mAndM.unmatched);
```